File: src/main/java/me/lokka30/levelledmobs/listeners/PlayerPortalEventListener.java
package me.lokka30.levelledmobs.listeners;

import me.lokka30.levelledmobs.LevelledMobs;
import me.lokka30.levelledmobs.util.Utils;
import org.bukkit.World;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerPortalEvent;
import org.bukkit.event.player.PlayerTeleportEvent;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.scheduler.BukkitRunnable;
import org.jetbrains.annotations.NotNull;

import java.util.ConcurrentModificationException;

public class PlayerPortalEventListener implements Listener {
    public PlayerPortalEventListener(final LevelledMobs main){
        this.main = main;
    }

    private final LevelledMobs main;

    @EventHandler(ignoreCancelled = true, priority = EventPriority.MONITOR)
    public void onPlayerPortalEvent(final @NotNull PlayerPortalEvent event){
        if (event.getCause() != PlayerTeleportEvent.TeleportCause.NETHER_PORTAL) return;
        if (event.getTo().getWorld() == null) return;
        final boolean isToNether = (event.getTo().getWorld().getEnvironment() == World.Environment.NETHER);

        final Player player = event.getPlayer();

        // store the player's portal coords in the nether.  only used for player levelling
        main.companion.setPlayerNetherPortalLocation(player, event.getTo());
        final String locationStr = String.format("%s,%s,%s,%s",
                event.getTo().getWorld().getName(), event.getTo().getBlockX(), event.getTo().getBlockY(), event.getTo().getBlockZ());

        final BukkitRunnable runnable = new BukkitRunnable() {
            @Override
            public void run() {
                if (isToNether)
                    main.companion.setPlayerNetherPortalLocation(player, player.getLocation());
                else
                    main.companion.setPlayerWorldPortalLocation(player, player.getLocation());

                try{
                    if (isToNether)
                        event.getPlayer().getPersistentDataContainer().set(main.namespaced_keys.playerNetherCoords, PersistentDataType.STRING, locationStr);
                    else
                        event.getPlayer().getPersistentDataContainer().set(main.namespaced_keys.playerNetherCoords_IntoWorld, PersistentDataType.STRING, locationStr);
                }
                catch (ConcurrentModificationException e){
                    Utils.logger.warning("Error updating PDC on " + player.getName() + ", " + e.getMessage());
                }
            }
        };

        // for some reason event#getTo has different coords that the actual nether portal
        // delay for 1 ticket and grab the player location instead
        runnable.runTaskLater(main, 1L);
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/listeners/PlayerPortalEventListener.kt
package io.github.arcaneplugins.levelledmobs.listeners

import io.github.arcaneplugins.levelledmobs.LevelledMobs
import io.github.arcaneplugins.levelledmobs.misc.NamespacedKeys
import io.github.arcaneplugins.levelledmobs.util.Log
import org.bukkit.World
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerPortalEvent
import org.bukkit.event.player.PlayerTeleportEvent
import org.bukkit.persistence.PersistentDataType
import org.bukkit.scheduler.BukkitRunnable

/**
 * Provides logic for when a player enters a portal.
 * Used for various spawn distance placeholders
 *
 * @author stumper66
 * @since 3.3.0
 */
class PlayerPortalEventListener : Listener {
    @EventHandler(ignoreCancelled = true, priority = EventPriority.MONITOR)
    fun onPlayerPortalEvent(event: PlayerPortalEvent) {
        if (event.cause != PlayerTeleportEvent.TeleportCause.NETHER_PORTAL) {
            return
        }
        if (event.to.world == null) {
            return
        }
        val isToNether = (event.to.world.environment
                == World.Environment.NETHER)

        val player = event.player
        val main = LevelledMobs.instance

        // store the player's portal coords in the nether.  only used for player levelling
        main.mainCompanion.setPlayerNetherPortalLocation(player, event.to)
        val locationStr = "${event.to.world.name},${event.to.blockX},${event.to.blockY},${event.to.blockZ}"

        val runnable: BukkitRunnable = object : BukkitRunnable() {
            override fun run() {
                if (isToNether) {
                    main.mainCompanion.setPlayerNetherPortalLocation(player, player.location)
                } else {
                    main.mainCompanion.setPlayerWorldPortalLocation(player, player.location)
                }

                try {
                    if (isToNether) {
                        event.player.persistentDataContainer
                            .set(
                                NamespacedKeys.playerNetherCoords, PersistentDataType.STRING,
                                locationStr
                            )
                    } else {
                        event.player.persistentDataContainer
                            .set(
                                NamespacedKeys.playerNetherCoordsIntoWorld,
                                PersistentDataType.STRING, locationStr
                            )
                    }
                } catch (e: ConcurrentModificationException) {
                    Log.war(
                        "Error updating PDC on ${player.name}, ${e.message}"
                    )
                }
            }
        }

        // for some reason event#getTo has different coords that the actual nether portal
        // delay for 1 ticket and grab the player location instead
        runnable.runTaskLater(main, 1L)
    }
}